pyccel/ast/core.py

```diff
@@ -1130,7 +1153,7 @@ def __init__(
         if not iterable(variables):
             raise TypeError('variables must be an iterable')
         for i in variables:
-            if not isinstance(i, Variable):
+            if not isinstance(i, (Variable, Assign)):
                 raise TypeError('Only a Variable instance is allowed.')
```

Comment:
I think you forget to mention in the error that the Assign instance is also allowed.

Reply:
Actually I left this on purpose but I'm not sure the review process is going to let me leave it like that ^^'. Normally you should provide a `Variable`. The **only** exception is the `initialised` boolean which needs to be initialised outside of the module initialisation function, but is not a `Constant`. In this one case there needs to be some way of declaring the variable with an initial value, and this was the best that I found. If you can think of another way to do it then I suspect that would be better

Reply:
I don't think I get your Idea, can you explain more.

Reply:
Currently the `init_function` is created [here](https://github.com/pyccel/pyccel/blob/devel-issue739/pyccel/parser/semantic.py#L1653-L1663)
It always has the same contents, which are:
```
def init_function():
    if initialised:
        # module_body
```
It would be possible to create the function in `Module.__init__` from module_body. However keeping the name clash avoidance would be harder and you would still need to collect the function afterwards to add it to the scope

Reply:
I think I found a better solution: 0b904013